stan/math/prim/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,239 @@
+#ifndef STAN_MATH_PRIM_MAT_VECTORIZE_APPLY_BINARY_SCALAR_HPP
+#define STAN_MATH_PRIM_MAT_VECTORIZE_APPLY_BINARY_SCALAR_HPP
+
+#include <stan/math/prim/mat/err/check_matching_dims.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <Eigen/Dense>
+#include <vector>
+
+namespace stan {
+
+  namespace math {
+
+    /**
+     * Base template class for vectorization of binary scalar functions
+     * defined by a template class <code>F</code> to a scalar,
+     * standard library vector, or Eigen dense matrix expression
+     * template.
+     *
+     * <p>The base class applies to any Eigen dense matrix expression
+     * template.  Specializations define applications to scalars
+     * (primitive or autodiff in the corresponding autodiff library
+     * directories) or to standard library vectors of vectorizable
+     * types (primitives, Eigen dense matrix expressions, or further
+     * standard vectors).
+     *
+     * <p>Each specialization must define the typedef
+     * <code>return_t</code> for the vectorized return type and the
+     * function <code>apply</code> which defines the vectorization or
+     * base application of the function defined statically by the
+     * class F.  The function definition class F defines a static
+     * function <code>fun()</code>, which defines the function's
+     * behavior on scalars.
+     *
+     * @tparam F Type of function to apply.
+     * @tparam T1 Type of first argument to which function is applied.
+     * @tparam T2 Type of second argument to which function is applied.
+     */
+    template <typename F, typename T1, typename T2>
+    struct apply_scalar_binary {
+      /**
+       * Type of underlying scalar for the matrix type T1.
```

Comment:
The doc goes with each variable --- up to you whether you want to doc simple typedefs, but this doc looks useful.